commit b94b50289622e816adc9f94111cfc2679c80177c
Author: Jesse Brandeburg <jesse.brandeburg@intel.com>
Date:   Tue Jan 19 14:15:59 2010 +0000

    e1000e: enhance frame fragment detection
    
    Originally patched by Neil Horman <nhorman@tuxdriver.com>
    
    e1000e could with a jumbo frame enabled interface, and packet split disabled,
    receive a packet that would overflow a single rx buffer.  While in practice
    very hard to craft a packet that could abuse this, it is possible.
    
    this is related to CVE-2009-4538
    
    Signed-off-by: Jesse Brandeburg <jesse.brandeburg@intel.com>
    CC: Neil Horman <nhorman@tuxdriver.com>
    Signed-off-by: Jeff Kirsher <jeffrey.t.kirsher@intel.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Backported to Debian's 2.6.24 by dann frazier <dannf@debian.org>

diff -urpN linux-source-2.6.24.orig/drivers/net/e1000e/e1000.h linux-source-2.6.24/drivers/net/e1000e/e1000.h
--- linux-source-2.6.24.orig/drivers/net/e1000e/e1000.h	2008-01-24 15:58:37.000000000 -0700
+++ linux-source-2.6.24/drivers/net/e1000e/e1000.h	2010-01-31 22:06:30.000000000 -0700
@@ -267,11 +267,13 @@ struct e1000_adapter {
 	unsigned long led_status;
 
 	unsigned int flags;
+	unsigned int flags2;
 };
 
 struct e1000_info {
 	enum e1000_mac_type	mac;
 	unsigned int		flags;
+	unsigned int		flags2;
 	u32			pba;
 	s32			(*get_invariants)(struct e1000_adapter *);
 	struct e1000_mac_operations *mac_ops;
@@ -311,6 +313,8 @@ struct e1000_info {
 #define FLAG_RX_CSUM_ENABLED              (1 << 28)
 #define FLAG_TSO_FORCE                    (1 << 29)
 
+#define FLAG2_IS_DISCARDING               (1 << 2)
+
 #define E1000_RX_DESC_PS(R, i)	    \
 	(&(((union e1000_rx_desc_packet_split *)((R).desc))[i]))
 #define E1000_GET_DESC(R, i, type)	(&(((struct type *)((R).desc))[i]))
diff -urpN linux-source-2.6.24.orig/drivers/net/e1000e/netdev.c linux-source-2.6.24/drivers/net/e1000e/netdev.c
--- linux-source-2.6.24.orig/drivers/net/e1000e/netdev.c	2009-11-04 18:42:03.000000000 -0700
+++ linux-source-2.6.24/drivers/net/e1000e/netdev.c	2010-01-31 22:06:30.000000000 -0700
@@ -389,14 +389,24 @@ static bool e1000_clean_rx_irq(struct e1
 
 		length = le16_to_cpu(rx_desc->length);
 
-		/* !EOP means multiple descriptors were used to store a single
-		 * packet, also make sure the frame isn't just CRC only */
-		if (!(status & E1000_RXD_STAT_EOP) || (length <= 4)) {
+		/*
+		 * !EOP means multiple descriptors were used to store a single
+		 * packet, if that's the case we need to toss it.  In fact, we
+		 * need to toss every packet with the EOP bit clear and the
+		 * next frame that _does_ have the EOP bit set, as it is by
+		 * definition only a frame fragment
+		 */
+		if (unlikely(!(status & E1000_RXD_STAT_EOP)))
+			adapter->flags2 |= FLAG2_IS_DISCARDING;
+
+		if (adapter->flags2 & FLAG2_IS_DISCARDING) {
 			/* All receives must fit into a single buffer */
 			ndev_dbg(netdev, "%s: Receive packet consumed "
 				 "multiple buffers\n", netdev->name);
 			/* recycle */
 			buffer_info->skb = skb;
+			if (status & E1000_RXD_STAT_EOP)
+				adapter->flags2 &= ~FLAG2_IS_DISCARDING;
 			goto next_desc;
 		}
 
@@ -648,10 +658,16 @@ static bool e1000_clean_rx_irq_ps(struct
 				 PCI_DMA_FROMDEVICE);
 		buffer_info->dma = 0;
 
-		if (!(staterr & E1000_RXD_STAT_EOP)) {
+		/* see !EOP comment in other rx routine */
+		if (!(staterr & E1000_RXD_STAT_EOP))
+			adapter->flags2 |= FLAG2_IS_DISCARDING;
+
+		if (adapter->flags2 & FLAG2_IS_DISCARDING) {
 			ndev_dbg(netdev, "%s: Packet Split buffers didn't pick "
 				 "up the full packet\n", netdev->name);
 			dev_kfree_skb_irq(skb);
+			if (staterr & E1000_RXD_STAT_EOP)
+				adapter->flags2 &= ~FLAG2_IS_DISCARDING;
 			goto next_desc;
 		}
 
@@ -813,6 +829,7 @@ static void e1000_clean_rx_ring(struct e
 
 	rx_ring->next_to_clean = 0;
 	rx_ring->next_to_use = 0;
+	adapter->flags2 &= ~FLAG2_IS_DISCARDING;
 
 	writel(0, adapter->hw.hw_addr + rx_ring->head);
 	writel(0, adapter->hw.hw_addr + rx_ring->tail);
@@ -3764,6 +3781,7 @@ static int __devinit e1000_probe(struct 
 	adapter->ei = ei;
 	adapter->pba = ei->pba;
 	adapter->flags = ei->flags;
+	adapter->flags2 = ei->flags2;
 	adapter->hw.adapter = adapter;
 	adapter->hw.mac.type = ei->mac;
 	adapter->msg_enable = (1 << NETIF_MSG_DRV | NETIF_MSG_PROBE) - 1;
